File: cs/Token.cs
namespace Morth;

public class Token
{
    public string Text { get; }
    public SourceLocation Location { get; }

    public Token(string text, SourceLocation location)
    {
        Text = text;
        Location = location;
    }

    public override string ToString()
    {
        return $"{Location}: '{Text}'";
    }
}

File: haskell/lib/Morth/Token.hs
module Morth.Token (Token (..), TokenKind (..), kindReadableName) where

import qualified Data.Text.Lazy as TL
import Morth.Location (Location (..))

data TokenKind
  = TokenWord TL.Text
  | TokenInt Int
  | TokenStr TL.Text
  | TokenChar Char
  deriving (Show)

kindReadableName :: TokenKind -> TL.Text
kindReadableName (TokenWord _) = "word"
kindReadableName (TokenInt _) = "integer"
kindReadableName (TokenStr _) = "string"
kindReadableName (TokenChar _) = "character"

data Token = Token
  { location :: Location
  , kind :: TokenKind
  }
  deriving (Show)
